File: benchmarks/python/load_feed.py
from dataclasses import dataclass

import httpx


@dataclass
class Feed:
    user_id: str
    posts: list[str]


@dataclass
class NetworkError:
    status: int


@dataclass
class DecodeError:
    reason: str


FeedError = NetworkError | DecodeError


@dataclass
class FeedOk:
    value: Feed


@dataclass
class FeedErr:
    error: FeedError


async def load_feed(user_id: str) -> FeedOk | FeedErr:
    async with httpx.AsyncClient() as client:
        try:
            user_resp = await client.get(f"/api/users/{user_id}")
            if user_resp.status_code >= 400:
                return FeedErr(NetworkError(status=user_resp.status_code))
            posts_resp = await client.get(f"/api/users/{user_id}/posts")
            if posts_resp.status_code >= 400:
                return FeedErr(NetworkError(status=posts_resp.status_code))
            posts_data = posts_resp.json()
            return FeedOk(Feed(user_id=user_id, posts=posts_data["titles"]))
        except Exception as e:
            return FeedErr(DecodeError(reason=str(e)))

File: benchmarks/rust/load_feed.rs
use reqwest::Client;

#[derive(Debug)]
pub struct Feed {
    pub user_id: String,
    pub posts: Vec<String>,
}

#[derive(Debug)]
pub enum FeedError {
    Network { status: u16 },
    Decode { reason: String },
}

pub async fn load_feed(client: &Client, user_id: &str) -> Result<Feed, FeedError> {
    let user_resp = client
        .get(format!("/api/users/{}", user_id))
        .send()
        .await
        .map_err(|e| FeedError::Decode { reason: e.to_string() })?;
    if !user_resp.status().is_success() {
        return Err(FeedError::Network { status: user_resp.status().as_u16() });
    }
    let posts_resp = client
        .get(format!("/api/users/{}/posts", user_id))
        .send()
        .await
        .map_err(|e| FeedError::Decode { reason: e.to_string() })?;
    if !posts_resp.status().is_success() {
        return Err(FeedError::Network { status: posts_resp.status().as_u16() });
    }
    #[derive(serde::Deserialize)]
    struct PostsData {
        titles: Vec<String>,
    }
    let posts_data: PostsData = posts_resp
        .json()
        .await
        .map_err(|e| FeedError::Decode { reason: e.to_string() })?;
    Ok(Feed { user_id: user_id.to_string(), posts: posts_data.titles })
}
